# Workgroup 7: Debiased Machine Learning

## Group 2: Diego Alonso Gómez and Partner

In [12]:
library(hdm)
library(xtable)
GrowthData <- GrowthData
save(GrowthData, file = "../data/GrowthData.RData")
growth <- GrowthData
attach(growth)
names(growth)

The following objects are masked from growth (pos = 3):

    bmp1l, ex1, fert1, fert65, freeop, freetar, gde1, gdpsh465,
    geerec1, geetot1, govsh41, govwb1, gpop1, gvxdxe41, h65, hf65,
    high65, highc65, highcf65, highcm65, highf65, highm65, hm65,
    human65, humanf65, humanm65, hyr65, hyrf65, hyrm65, im1, intercept,
    invsh41, lifee065, mort1, mort65, no65, nof65, nom65, Outcome, p65,
    pf65, pinstab1, pm65, pop1565, pop65, pop6565, s65, sec65, secc65,
    seccf65, seccm65, secf65, secm65, sf65, sm65, syr65, syrf65,
    syrm65, teapri65, teasec65, tot1, worker65, xr65



[1] "Outcome"   "intercept" "gdpsh465"  "bmp1l"     "freeop"    "freetar"  
 [7] "h65"       "hm65"      "hf65"      "p65"       "pm65"      "pf65"     
[13] "s65"       "sm65"      "sf65"      "fert65"    "mort65"    "lifee065" 
[19] "gpop1"     "fert1"     "mort1"     "invsh41"   "geetot1"   "geerec1"  
[25] "gde1"      "govwb1"    "govsh41"   "gvxdxe41"  "high65"    "highm65"  
[31] "highf65"   "highc65"   "highcm65"  "highcf65"  "human65"   "humanm65" 
[37] "humanf65"  "hyr65"     "hyrm65"    "hyrf65"    "no65"      "nom65"    
[43] "nof65"     "pinstab1"  "pop65"     "worker65"  "pop1565"   "pop6565"  
[49] "sec65"     "secm65"    "secf65"    "secc65"    "seccm65"   "seccf65"  
[55] "syr65"     "syrm65"    "syrf65"    "teapri65"  "teasec65"  "ex1"      
[61] "im1"       "xr65"      "tot1"

In [13]:
head(growth)

Outcome     intercept gdpsh465 bmp1l  freeop   freetar  h65   hm65  hf65 
1 -0.02433575 1         6.591674 0.2837 0.153491 0.043888 0.007 0.013 0.001
2  0.10047257 1         6.829794 0.6141 0.313509 0.061827 0.019 0.032 0.007
3  0.06705148 1         8.895082 0.0000 0.204244 0.009186 0.260 0.325 0.201
4  0.06408917 1         7.565275 0.1997 0.248714 0.036270 0.061 0.070 0.051
5  0.02792955 1         7.162397 0.1740 0.299252 0.037367 0.017 0.027 0.007
6  0.04640744 1         7.218910 0.0000 0.258865 0.020880 0.023 0.038 0.006
  p65  ... seccf65 syr65 syrm65 syrf65 teapri65 teasec65 ex1    im1    xr65 
1 0.29 ...  0.04   0.033 0.057  0.010  47.6     17.3     0.0729 0.0667 0.348
2 0.91 ...  0.64   0.173 0.274  0.067  57.1     18.0     0.0940 0.1438 0.525
3 1.00 ... 18.14   2.573 2.478  2.667  26.5     20.7     0.1741 0.1750 1.082
4 1.00 ...  2.63   0.438 0.453  0.424  27.8     22.7     0.1265 0.1496 6.625
5 0.82 ...  2.11   0.257 0.287  0.229  34.5     17.6     0.1211 0.1308 2.500
6 0.50 ...  1.46   0.160 0.174  0.146  34.3      8.1     0.0634 0.0762 1.000
  tot1     
1 -0.014727
2  0.005750
3 -0.010040
4 -0.002195
5  0.003283
6 -0.001747

In [24]:
dim(growth)

[1] 90 63

In [25]:
colnames(growth)[3:63]

[1] "gdpsh465" "bmp1l"    "freeop"   "freetar"  "h65"      "hm65"    
 [7] "hf65"     "p65"      "pm65"     "pf65"     "s65"      "sm65"    
[13] "sf65"     "fert65"   "mort65"   "lifee065" "gpop1"    "fert1"   
[19] "mort1"    "invsh41"  "geetot1"  "geerec1"  "gde1"     "govwb1"  
[25] "govsh41"  "gvxdxe41" "high65"   "highm65"  "highf65"  "highc65" 
[31] "highcm65" "highcf65" "human65"  "humanm65" "humanf65" "hyr65"   
[37] "hyrm65"   "hyrf65"   "no65"     "nom65"    "nof65"    "pinstab1"
[43] "pop65"    "worker65" "pop1565"  "pop6565"  "sec65"    "secm65"  
[49] "secf65"   "secc65"   "seccm65"  "seccf65"  "syr65"    "syrm65"  
[55] "syrf65"   "teapri65" "teasec65" "ex1"      "im1"      "xr65"    
[61] "tot1"

### Preprocessing

In [35]:
#################################  Find Variable Names from Dataset ########################

varlist <- function (df=NULL,type=c("numeric","factor","character"), pattern="", exclude=NULL) {
  vars <- character(0)
  if (any(type %in% "numeric")) {
    vars <- c(vars,names(df)[sapply(df,is.numeric)])
  }
  if (any(type %in% "factor")) {
    vars <- c(vars,names(df)[sapply(df,is.factor)])
  }  
  if (any(type %in% "character")) {
    vars <- c(vars,names(df)[sapply(df,is.character)])
  }  
  vars[(!vars %in% exclude) & grepl(vars,pattern=pattern)]
}

################################# Create Variables ####################################################


################################ Variables ##################################
# Treatment Variable
d     <- "gdpsh465"    # THE INICIAL WEALTH

# Outcome Variable
y     <- "Outcome"

# Other Control Variables
X1    <-   c(colnames(growth)[3:63])  

#################################  Partial out Fixed Effects ########################


# Variables to be Partialled-out
varlist <- c(y, d,X1)


Now, we can construct the treatment variable, the outcome variable and the matrix $Z$ that includes the control variables.

In [69]:
# Treatment Variable
D     <- growth["gdpsh465"]

# Outcome Variable
Y     <- growth["Outcome"]

# Construct matrix Z

Z <- growth[c(colnames(growth)[3:63])]

In [72]:
# To use in Python
save(data, file = "../data/growth_group2.RData")

In [74]:
# install.packages("lfe")

In [46]:
library(Matrix)

In [47]:
library(lfe)

Warning message:
"package 'lfe' was built under R version 3.6.3"

## y = outcome: growth rate  (Outcome)

## d = treatment: initial wealth  (gdpsh465)

## x = controls: country characteristics

The target parameter $\beta_1$ is the speed of convergence, which measures the speed at which poor countries catch up with rich countries. The relationship is captured by $\beta_1$, the *speed of convergence/divergence*, which measures the speed at which poor countries catch up $(\beta_1< 0)$ or fall behind $(\beta_1> 0)$ rich countries, after controlling for $W$. Our inference question here is: do poor countries grow faster than rich countries, controlling for educational and other characteristics? In other words, is the speed of convergence negative: $ \beta_1 <0?$ This is the Convergence Hypothesis predicted by the Solow Growth Model.

## 1° OLS without including the country characteristics.

After preprocessing the data, we first look at simple regression of $Y_{j,t}$ on $D_{j,t-1}$ without controls as a baseline model.

In [57]:
#baseline_formula <- as.formula(paste(y, "~", d ))
#baseline.ols <- lm(baseline_formula,data=rdata)

baseline.ols <- felm(Outcome ~ gdpsh465,data=growth)
est_baseline <- summary(baseline.ols)$coef[2,]
confint(baseline.ols)[2,]
est_baseline

2.5 %      97.5 % 
-0.01081013  0.01344355

Estimate  Std. Error     t value    Pr(>|t|) 
0.001316713 0.006102200 0.215776701 0.829661165

In [64]:
summary(baseline.ols)$coef

Estimate    Std. Error t value   Pr(>|t|) 
(Intercept) 0.035206978 0.04731822 0.7440470 0.4588302
gdpsh465    0.001316713 0.00610220 0.2157767 0.8296612

In [60]:
confint(baseline.ols)[2,]

2.5 %      97.5 % 
-0.01081013  0.01344355

The point estimate is $0.001316713$ with the confidence interval ranging from $-0.0108101278829859$ to $0.013443553109878$. This suggests that poor countries fall behind rich countries, so the convergence hypothesis predicted by the Solow Growth Model would not apply here. However, this first estimation provides us with a non-significant estimation, that due , and does not allow us to answer the question
about the convergence hypothesis since the confidence interval crosses 0 on the right side of the interval.

## 2° OLS including the country characteristics. (OLS with controls)

In [63]:
summary(control.ols)$coef

Estimate      Std. Error   t value     Pr(>|t|)   
intercept  0.247160893  0.78450163    0.31505466 0.755056170
gdpsh465  -0.009377989  0.02988773   -0.31377391 0.756018518
bmp1l     -0.068862679  0.03253065   -2.11685513 0.043289718
freeop     0.080068974  0.20786400    0.38519885 0.703000838
freetar   -0.488962605  0.41816285   -1.16931143 0.252136477
h65       -2.362098638  0.85729167   -2.75530338 0.010192435
hm65       0.707143400  0.52314511    1.35171560 0.187285919
hf65       1.693448425  0.50318881    3.36543337 0.002232683
p65        0.265526695  0.16429407    1.61616729 0.117271229
pm65       0.136952626  0.15121749    0.90566657 0.372840111
pf65      -0.331266948  0.16512299   -2.00618303 0.054583866
s65        0.039079306  0.18552248    0.21064459 0.834690082
sm65      -0.030668507  0.11679396   -0.26258642 0.794791807
sf65      -0.179917280  0.11809892   -1.52344556 0.138863011
fert65     0.006880834  0.02705142    0.25436128 0.801075252
mort65    -0.233454503  0.81742198   -0.28559852 0.777287824
lifee065  -0.014914473  0.19333559   -0.07714292 0.939058459
gpop1      0.970184579  1.81220835    0.53536039 0.596627593
fert1      0.008838189  0.03504148    0.25222077 0.802712696
mort1      0.066562928  0.68481002    0.09719911 0.923260701
invsh41    0.074461126  0.10844692    0.68661354 0.497973611
geetot1   -0.715104906  1.68013567   -0.42562331 0.673636102
geerec1    0.630004630  2.44737848    0.25742019 0.798736842
gde1      -0.443575219  1.67120674   -0.26542211 0.792628757
govwb1     0.337452364  0.43798695    0.77046214 0.447478336
govsh41    0.463174632  1.92543733    0.24055555 0.811652246
gvxdxe41  -0.793381582  2.05939737   -0.38524939 0.702963815
high65    -0.752483828  0.90572159   -0.83081141 0.413108816
highm65   -0.390261804  0.68124930   -0.57286195 0.571310302
highf65   -0.417748705  0.56149048   -0.74399963 0.463075062
...       ...           ...          ...         ...        
highcf65   3.920771e-01 7.660246e-01  0.51183360 0.61278088 
human65    2.337269e+00 3.307269e+00  0.70670678 0.48558873 
humanm65  -1.209249e+00 1.618516e+00 -0.74713475 0.46121072 
humanf65  -1.103947e+00 1.684693e+00 -0.65528075 0.51763499 
hyr65      5.491388e+01 2.388729e+01  2.29887430 0.02918450 
hyrm65     1.293505e+01 2.317142e+01  0.55823278 0.58112223 
hyrf65     9.092580e+00 1.766958e+01  0.51458948 0.61087820 
no65       3.720986e-02 1.319733e-01  0.28194994 0.78005541 
nom65     -2.119769e-02 6.495975e-02 -0.32632032 0.74660861 
nof65     -1.685782e-02 6.700258e-02 -0.25159954 0.80318810 
pinstab1  -4.997113e-02 3.092081e-02 -1.61610019 0.11728579 
pop65      1.031796e-07 1.317950e-07  0.78287929 0.44026959 
worker65   3.407897e-02 1.561909e-01  0.21818786 0.82886516 
pop1565   -4.655346e-01 4.713340e-01 -0.98769582 0.33176044 
pop6565   -1.357450e+00 6.349424e-01 -2.13791049 0.04139326 
sec65     -1.089278e-02 3.076618e-01 -0.03540506 0.97200796 
secm65     3.343664e-03 1.511923e-01  0.02211530 0.98251284 
secf65    -2.304329e-03 1.579715e-01 -0.01458699 0.98846512 
secc65    -4.915284e-01 7.290408e-01 -0.67421246 0.50570489 
seccm65    2.596016e-01 3.556533e-01  0.72992873 0.47149686 
seccf65    2.206524e-01 3.733333e-01  0.59103325 0.55923950 
syr65     -7.555807e-01 7.976763e+00 -0.09472273 0.92520970 
syrm65     3.109007e-01 3.896731e+00  0.07978502 0.93697576 
syrf65     7.592811e-01 4.110634e+00  0.18471143 0.85478676 
teapri65   3.954523e-05 7.700406e-04  0.05135474 0.95940747 
teasec65   2.496674e-04 1.171230e-03  0.21316678 0.83274136 
ex1       -5.804081e-01 2.418472e-01 -2.39989593 0.02329485 
im1        5.914450e-01 2.502978e-01  2.36296558 0.02530892 
xr65      -1.037768e-04 5.416760e-05 -1.91584652 0.06564690 
tot1      -1.279011e-01 1.125951e-01 -1.13593842 0.26561200

In [62]:
control.ols <- lm(Outcome~.-1,data=growth)
est_ols <- summary(control.ols)$coef[2,]
confint(control.ols)[2,]
est_ols

2.5 %      97.5 % 
-0.07060022  0.05184424

Estimate   Std. Error      t value     Pr(>|t|) 
-0.009377989  0.029887726 -0.313773911  0.756018518

After controlling The point estimate is $-0.00937798878258103$ with the confidence interval ranging from $
-0.0706002209228739$ to $
-0.0706002209228739$. This can suggest that poor countries catch up rich countries, so the convergence hypothesis predicted by the Solow Growth Model would be accepted here. However, this estimation also provides us with a non-significant estimation, that due , and does not allow us to answer the question
about the convergence hypothesis since the confidence interval crosses 0 on the right side of the interval.

## 3° DML using Lasso to predict y an d.

# DML algorithm

Here we perform inference of the predictive coefficient $\beta$ in our partially linear statistical model, 

$$
Y = D\beta + g(Z) + \epsilon, \quad E (\epsilon | D, Z) = 0,
$$

using the **double machine learning** approach. 

For $\tilde Y = Y- E(Y|Z)$ and $\tilde D= D- E(D|Z)$, we can write
$$
\tilde Y = \alpha \tilde D + \epsilon, \quad E (\epsilon |\tilde D) =0.
$$

Using cross-fitting, we employ modern regression methods
to build estimators $\hat \ell(Z)$ and $\hat m(Z)$ of $\ell(Z):=E(Y|Z)$ and $m(Z):=E(D|Z)$ to obtain the estimates of the residualized quantities:

$$
\tilde Y_i = Y_i  - \hat \ell (Z_i),   \quad \tilde D_i = D_i - \hat m(Z_i), \quad \text{ for each } i = 1,\dots,n.
$$

Finally, using ordinary least squares of $\tilde Y_i$ on $\tilde D_i$, we obtain the 
estimate of $\beta$.

The following algorithm comsumes $Y, D, Z$, and a machine learning method for learning the residuals $\tilde Y$ and $\tilde D$, where the residuals are obtained by cross-validation (cross-fitting). Then, it prints the estimated coefficient $\beta$ and the corresponding standard error from the final OLS regression.

In [76]:
DML2.for.PLM <- function(z, d, y, dreg, yreg, nfold=2) {
  nobs <- nrow(z) #number of observations
  foldid <- rep.int(1:nfold,times = ceiling(nobs/nfold))[sample.int(nobs)] #define folds indices
  I <- split(1:nobs, foldid)  #split observation indices into folds  
  ytil <- dtil <- rep(NA, nobs)
  cat("fold: ")
  for(b in 1:length(I)){
    dfit <- dreg(z[-I[[b]],], d[-I[[b]]]) #take a fold out
    yfit <- yreg(z[-I[[b]],], y[-I[[b]]]) # take a foldt out
    dhat <- predict(dfit, z[I[[b]],], type="response") #predict the left-out fold 
    yhat <- predict(yfit, z[I[[b]],], type="response") #predict the left-out fold  
    dtil[I[[b]]] <- (d[I[[b]]] - dhat) #record residual for the left-out fold
    ytil[I[[b]]] <- (y[I[[b]]] - yhat) #record residial for the left-out fold
    cat(b," ")
        }
  #rfit <- lm(ytil ~ dtil)    #estimate the main parameter by regressing one residual on the other
  data <- data.frame(cbind(ytil, dtil))
  rfit <- felm(ytil ~ dtil,data=growth) 
  coef.est <- coef(rfit)[2]  #extract coefficient
  #HC <- vcovHC(rfit)
  se    <- summary(rfit,robust=T)$coefficients[2,2] #record robust standard error by County
  cat(sprintf("\ncoef (se) = %g (%g)\n", coef.est , se))  #printing output
  return( list(coef.est =coef.est , se=se, dtil=dtil, ytil=ytil, rfit=rfit) ) #save output and residuals 
}

Now, we apply the Double Machine Learning (DML) approach with different machine learning methods. First, we load the relevant libraries.

In [77]:
# install.packages("hdm")

Warning message:
"package 'hdm' is in use and will not be installed"

In [78]:
library(foreach)
library(hdm)
library(glmnet)
library(sandwich)
library(randomForest)

Loaded glmnet 2.0-16

randomForest 4.6-14
Type rfNews() to see new features/changes/bug fixes.


Let us, construct the input matrices.

In [102]:
# Treatment Variable
D     <- growth["gdpsh465"]

# Outcome Variable
Y     <- growth["Outcome"]

# Construct matrix Z

Z <- growth[c(colnames(growth)[4:63])]

In [103]:
y <- as.matrix(Y)

In [104]:
d <- as.matrix(D)

In [105]:
z <- as.matrix(Z)

In [106]:
head(data.frame(cbind(y,d)))

Outcome     gdpsh465
1 -0.02433575 6.591674
2  0.10047257 6.829794
3  0.06705148 8.895082
4  0.06408917 7.565275
5  0.02792955 7.162397
6  0.04640744 7.218910

In the following, we apply the DML approach with the differnt versions of lasso.


## 3° DML using Lasso to predict y an d.

In [107]:
#DML with Lasso:
set.seed(123)
dreg <- function(z,d){ rlasso(z,d, post=FALSE) } #ML method= lasso from hdm 
yreg <- function(z,y){ rlasso(z,y, post=FALSE) } #ML method = lasso from hdm

In [108]:
DML2.lasso = DML2.for.PLM(z, d, y, dreg, yreg, nfold=10)

fold: 1  2  3  4  5  6  7  8  9  10  
coef (se) = -0.0370317 (0.0161774)


Here we have that to estimate the $\hat{\beta}$ related to the effect of $D$ on $Y$ was constructed as a solution of the following penalized least square problem:
$$
        min_{b \in R^p} \sum_i (Y_i - b^\prime X_i )^2 + \lambda \dot \sum_{j=1}^{p} |b_j|
$$

So, DML with lasso give us a estimate parameter equal to $-0.0370317$, with a standard deviation of $0.0161774$. This beta estimate suggest that poor countries catch up rich countries, so the convergence hypothesis predicted by the Solow Growth Model would be accepted here. Besides that, we can see that this standard deviation is lower than the other two previous ones we got.

## 4° DML using Post-Lasso to predict y an d.

In [110]:
#DML with Post-Lasso:
dreg <- function(z,d){ rlasso(z,d, post=T) } #ML method= lasso from hdm 
yreg <- function(z,y){ rlasso(z,y, post=T) } #ML method = lasso from hdm
DML2.post = DML2.for.PLM(z, d, y, dreg, yreg, nfold=10)

fold: 1  2  3  4  5  6  7  8  9  10  
coef (se) = -0.0368285 (0.0141174)


Here we have that to estimate the $\hat{\beta}$ related to the effect of $D$ on $Y$ it was used the Lasso-selected set of regressors to refit the model by least squares. This method is callend the "least squares post Lasso" and was constructed as a solution of the following penalized least square problem.

So, this technique applied with DML give us a estimate parameter equal to $-0.0368285$, with a standard deviation of $0.0141174$. This beta estimate suggest that poor countries catch up rich countries, so the convergence hypothesis predicted by the Solow Growth Model would also be accepted here. Besides that, we can see that this standard deviation is even lower than the one we got with DML- Lasso.

## 5° DML using Random Forest to predict y an d.

In [113]:
#DML with Random Forest:
dreg <- function(z,d){ randomForest(z, d) } #ML method=Forest 
yreg <- function(z,y){ randomForest(z, y) } #ML method=Forest
set.seed(1)
DML2.RF = DML2.for.PLM(z, d, y, dreg, yreg, nfold=2) # set to 2 due to computation time

fold: 1  2  
coef (se) = -0.0290514 (0.0128382)


Here we have that to estimate the $\hat{\beta}$ related to the effect of $D$ on $Y$ it was used DML with Random forest to predict Y and D. 
So, this technique give us a estimate parameter equal to $-0.0290514$, with a standard deviation of $0.0128382$. This beta estimate also suggest that poor countries catch up rich countries, so the convergence hypothesis predicted by the Solow Growth Model would also be accepted here. Besides that, we can see that this standard deviation is even lower than the one we got with DML- Post Lasso.

## 6° Run the best method i.e. the best combination of methods to predict y and d. 

To be able to find the best method, we will execute the DML algorithm with other machine learning techniques:

In [122]:
#DML with cross-validated Lasso:
dreg <- function(z,d){ cv.glmnet(z,d,family="gaussian", alpha=1) } #ML method = cvlasso from glmnet 
yreg <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=1) }  #ML method = cvlasso from glmnet 
DML2.lasso.cv = DML2.for.PLM(z, d, y, dreg, yreg, nfold=5)

dreg <- function(z,d){ cv.glmnet(z,d,family="gaussian", alpha=0.5) } #ML method = elastic net from glmnet 
yreg <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=0.5) }  #ML method = elastic net from glmnet 
DML2.elnet = DML2.for.PLM(z, d, y, dreg, yreg, nfold=5)

dreg <- function(z,d){ cv.glmnet(z,d,family="gaussian", alpha=0) } #ML method = ridge from glmnet 
yreg <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=0) }  #ML method = ridge from glmnet 
DML2.ridge = DML2.for.PLM(z, d, y, dreg, yreg, nfold=5)

fold: 1  2  3  4  5  
coef (se) = -0.0360247 (0.0153477)
fold: 1  2  3  4  5  
coef (se) = -0.0365546 (0.015524)
fold: 1  2  3  4  5  
coef (se) = -0.0327663 (0.0157992)


In [123]:
# DML with OLS used as the ML method
dreg <- function(z,d){  glmnet(z,d,family="gaussian", lambda=0) } #ML method = ols from glmnet 
yreg <- function(z,y){  glmnet(z,y,family="gaussian", lambda=0) }  #ML method = ols from glmnet 
DML2.ols = DML2.for.PLM(z, d, y, dreg, yreg, nfold=10)

fold: 1  2  3  4  5  6  7  8  9  10  
coef (se) = 0.0265087 (0.0134078)


Finally, let's see which method is actually better. We compute RMSE for predicting D and Y, and see which
of the methods works better.

In [134]:
mods<- list(DML2.ols, DML2.lasso, DML2.post, DML2.lasso.cv, DML2.ridge, DML2.elnet, DML2.RF)

RMSE.mdl<- function(mdl) {
RMSEY <- sqrt(mean(mdl$ytil)^2) 
RMSED <- sqrt(mean(mdl$dtil)^2) 
return( list(RMSEY=RMSEY, RMSED=RMSED))
}

#RMSE.mdl(DML2.lasso)

#DML2.lasso$ytil

Res<- lapply(mods, RMSE.mdl)


prRes.Y<- c( Res[[1]]$RMSEY,Res[[2]]$RMSEY, Res[[3]]$RMSEY, Res[[4]]$RMSEY, Res[[5]]$RMSEY,  Res[[6]]$RMSEY, Res[[7]]$RMSEY)
prRes.D<- c( Res[[1]]$RMSED,Res[[2]]$RMSED, Res[[3]]$RMSED, Res[[4]]$RMSED, Res[[5]]$RMSED, Res[[6]]$RMSED, Res[[7]]$RMSED)

prRes<- rbind(prRes.Y, prRes.D); 
rownames(prRes)<- c("RMSE D", "RMSE Y");
colnames(prRes)<- c("OLS", "Lasso", "Post-Lasso", "CV Lasso", "CV Ridge", "CV Elnet", "RF")
print(prRes,digit=6)

               OLS       Lasso  Post-Lasso    CV Lasso    CV Ridge    CV Elnet
RMSE D 0.000449556 0.000831523 0.000333837 8.08041e-19 0.000154909 0.000813289
RMSE Y 0.010418259 0.012439826 0.000541272 3.23386e-03 0.008698366 0.006694750
                RF
RMSE D 0.000844267
RMSE Y 0.029621814


Here we can see that the best method for predicting D is "CV Lasso", and the best method for predicting Y is also "CV Lasso".So, we'll run the algorithm following these two machine-learning methods for $D$ and $Y$: 

In [135]:
dreg_b <- function(z,d){ cv.glmnet(z,d,family="gaussian", alpha=1) } #ML method = cvlasso from glmnet for d 
yreg_b <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=1) }  #ML method = cvlasso from glmnet y
DML2.best= DML2.for.PLM(z, d, y, dreg_b, yreg_b, nfold=10)

fold: 1  2  3  4  5  6  7  8  9  10  
coef (se) = -0.0340352 (0.0160492)


## 7° Show your results in a table as we did in the lab.

Let's organize the results in a table.

In [138]:
library(xtable)

table <- matrix(0,9,2)
table[1,1] <- as.numeric(est_baseline[1])
table[2,1] <- as.numeric(est_ols[1])
table[3,1]   <- as.numeric(DML2.lasso$coef.est)
table[4,1]   <- as.numeric(DML2.post$coef.est)
table[5,1]  <-as.numeric(DML2.lasso.cv$coef.est)
table[6,1] <-as.numeric(DML2.elnet$coef.est)
table[7,1] <-as.numeric(DML2.ridge$coef.est)
table[8,1] <-as.numeric(DML2.RF$coef.est)
table[9,1] <-as.numeric(DML2.best$coef.est)
table[1,2] <- as.numeric(est_baseline[2])
table[2,2] <- as.numeric(est_ols[2])
table[3,2]   <- as.numeric(DML2.lasso$se)
table[4,2]   <- as.numeric(DML2.post$se)
table[5,2]  <-as.numeric(DML2.lasso.cv$se)
table[6,2] <-as.numeric(DML2.elnet$se)
table[7,2] <-as.numeric(DML2.ridge$se)
table[8,2] <-as.numeric(DML2.RF$se)
table[9,2] <-as.numeric(DML2.best$se)


################################# Print Results #################################

colnames(table) <- c("Estimate","Standard Error")
rownames(table) <- c("Baseline OLS", "Least Squares with controls", "Lasso", "Post-Lasso", "CV Lasso","CV Elnet", "CV Ridge", "Random Forest", 
                     "Best")

table

Estimate     Standard Error
Baseline OLS                 0.001316713 0.00610220    
Least Squares with controls -0.009377989 0.02988773    
Lasso                       -0.037031673 0.01617736    
Post-Lasso                  -0.036828539 0.01411740    
CV Lasso                    -0.036024675 0.01534774    
CV Elnet                    -0.036554650 0.01552403    
CV Ridge                    -0.032766296 0.01579923    
Random Forest               -0.029051448 0.01283821    
Best                        -0.034035162 0.01604917

In [139]:
print(table, digit=3)

                            Estimate Standard Error
Baseline OLS                 0.00132         0.0061
Least Squares with controls -0.00938         0.0299
Lasso                       -0.03703         0.0162
Post-Lasso                  -0.03683         0.0141
CV Lasso                    -0.03602         0.0153
CV Elnet                    -0.03655         0.0155
CV Ridge                    -0.03277         0.0158
Random Forest               -0.02905         0.0128
Best                        -0.03404         0.0160
